File: lib/factorix/save_file.rb
# frozen_string_literal: true

require "zip"
require "zlib"

module Factorix
  SaveFile = Data.define(:version, :mods, :startup_settings)

  # Data structure for Factorio save file information
  #
  # SaveFile provides functionality to extract MOD information and startup settings
  # from Factorio save files (.zip format containing level.dat0 or level-init.dat).
  #
  # @!attribute [r] version
  #   @return [Factorix::GameVersion] Game version from the save file
  # @!attribute [r] mods
  #   @return [Hash<String, Factorix::MODState>] Hash of MOD name to MODState
  # @!attribute [r] startup_settings
  #   @return [Factorix::MODSettings::Section] Startup settings section
  class SaveFile
    # Level file names to search for, in priority order
    LEVEL_FILE_NAMES = %w[level.dat0 level-init.dat].freeze
    private_constant :LEVEL_FILE_NAMES

    # Load a save file and extract MOD information and settings
    #
    # @param path [Pathname] Path to the save file (.zip)
    # @return [SaveFile] Extracted save file data
    # @raise [Errno::ENOENT] If save file or level file not found
    # @raise [Factorix::Error] If save file format is invalid
    def self.load(path) = Parser.new(path).parse

    # Internal parser for save files
    class Parser
      # Initialize a new Parser instance
      #
      # @param path [Pathname] Path to the save file
      def initialize(path) = @path = path

      # Parse the save file and return extracted data
      #
      # @return [SaveFile] Extracted save file data
      # @raise [Errno::ENOENT] If save file or level file not found
      def parse
        open_level_file do |stream|
          deserializer = SerDes::Deserializer.new(stream)
          parse_save_header(deserializer)
          skip_unknown_bytes(deserializer)
          parse_startup_settings(deserializer)
        end

        SaveFile.new(version: @version, mods: @mods, startup_settings: @startup_settings)
      end

      private def open_level_file
        Zip::File.open(@path) do |zip_file|
          LEVEL_FILE_NAMES.each do |file_name|
            entry = find_level_entry(zip_file, file_name)
            next unless entry

            stream = entry.get_input_stream
            return yield decompress_if_needed(stream)
          end

          raise Errno::ENOENT, "level.dat0 or level-init.dat not found in #{@path}"
        end
      end

      # Find a level file entry in the zip file
      #
      # @param zip_file [Zip::File] The zip file
      # @param file_name [String] The level file name to search for
      # @return [Zip::Entry, nil] The entry if found, nil otherwise
      private def find_level_entry(zip_file, file_name) = zip_file.glob("*/#{file_name}").first

      # Decompress stream if it's zlib compressed
      #
      # @param stream [IO] The input stream
      # @return [IO] Decompressed stream or original stream
      private def decompress_if_needed(stream)
        # Read CMF (Compression Method and Flags) byte
        cmf = stream.read(1)
        return StringIO.new("") if cmf.nil?

        stream.rewind

        # CMF = 0x78 indicates zlib compression (DEFLATE with 32K window)
        if cmf.unpack1("C") == 0x78
          StringIO.new(Zlib::Inflate.inflate(stream.read))
        else
          stream
        end
      end

      # Parse save header to extract game version and MOD list
      #
      # @param deserializer [Factorix::SerDes::Deserializer] The deserializer
      # @return [void]
      private def parse_save_header(deserializer)
        # Read game version
        @version = deserializer.read_game_version

        # Skip 1 byte after version
        deserializer.read_u8

        # Skip fields we don't need
        deserializer.read_str  # campaign
        deserializer.read_str  # level_name
        deserializer.read_str  # base_mod
        deserializer.read_u8   # difficulty
        deserializer.read_bool # finished
        deserializer.read_bool # player_won
        deserializer.read_str  # next_level
        deserializer.read_bool # can_continue
        deserializer.read_bool # finished_but_continuing
        deserializer.read_bool # saving_replay
        deserializer.read_bool # allow_non_admin_debug_options
        deserializer.read_mod_version # loaded_from (MODVersion, not GameVersion)
        deserializer.read_u16 # loaded_from_build
        deserializer.read_u8 # allowed_commands

        # Additional fields before the MOD list
        # These fields' purposes are not yet fully understood
        deserializer.read_bool  # Unknown boolean field
        deserializer.read_u32   # Unknown u32 field
        deserializer.read_bool  # Unknown boolean field

        # Read MOD list
        parse_mods(deserializer)
      end

      # Parse MOD list from save header
      #
      # @param deserializer [Factorix::SerDes::Deserializer] The deserializer
      # @return [void]
      private def parse_mods(deserializer)
        mods_count = deserializer.read_optim_u32
        @mods = {}

        mods_count.times do
          name = deserializer.read_str
          version = deserializer.read_mod_version
          _crc = deserializer.read_u32

          # All MODs in save file are enabled
          @mods[name] = MODState.new(enabled: true, version:)
        end
      end

      # Skip unknown 4 bytes after MOD list
      #
      # @param deserializer [Factorix::SerDes::Deserializer] The deserializer
      # @return [void]
      private def skip_unknown_bytes(deserializer) = deserializer.read_bytes(4)

      # Parse startup settings from save file
      #
      # @param deserializer [Factorix::SerDes::Deserializer] The deserializer
      # @return [void]
      private def parse_startup_settings(deserializer)
        raw_settings = deserializer.read_property_tree

        # Create a new Section and populate it
        @startup_settings = MODSettings::Section.new("startup")

        return unless raw_settings.is_a?(Hash)

        raw_settings.each do |key, value_hash|
          # Extract the actual value from the {"value" => X} hash
          next unless value_hash.is_a?(Hash)

          @startup_settings[key] = value_hash["value"]
        end
      end
    end
    private_constant :Parser
  end
end

File: internal/save/save_file.go
// Package save extracts MOD information and startup settings from Factorio
// save files (.zip archives containing level.dat0 or level-init.dat).
//
// This lives outside internal/mod because it needs internal/serdes, which
// itself imports internal/mod for the version types.
package save

import (
	"archive/zip"
	"bufio"
	"compress/zlib"
	"io"
	"strings"

	"github.com/sakuro/factorix/internal/mod"
	"github.com/sakuro/factorix/internal/serdes"
	"github.com/sakuro/factorix/internal/settings"
)

// Level file names to search for, in priority order.
var levelFileNames = []string{"level.dat0", "level-init.dat"}

// File is the information extracted from a save file.
type File struct {
	Version         mod.GameVersion
	MODs            []MODEntry // in save-file order; every recorded MOD is enabled
	StartupSettings *settings.Section
}

// MODEntry is one MOD recorded in the save header.
type MODEntry struct {
	Name    string
	Version mod.MODVersion
}

// Load extracts MOD information and startup settings from a save file.
func Load(path string) (*File, error) {
	zr, err := zip.OpenReader(path)
	if err != nil {
		return nil, &mod.FileFormatError{Path: path, Msg: "invalid zip file: " + err.Error()}
	}
	defer zr.Close()

	for _, fileName := range levelFileNames {
		entry := findLevelEntry(&zr.Reader, fileName)
		if entry == nil {
			continue
		}
		rc, err := entry.Open()
		if err != nil {
			return nil, &mod.FileFormatError{Path: path, Msg: err.Error()}
		}
		defer rc.Close()

		r, err := decompressIfNeeded(rc)
		if err != nil {
			return nil, &mod.FileFormatError{Path: path, Msg: err.Error()}
		}
		return parse(r)
	}
	return nil, &mod.FileFormatError{Path: path, Msg: "level.dat0 or level-init.dat not found"}
}

// The level file sits in the single top-level directory named after the save.
func findLevelEntry(zr *zip.Reader, fileName string) *zip.File {
	for _, f := range zr.File {
		parts := strings.Split(f.Name, "/")
		if len(parts) == 2 && parts[1] == fileName {
			return f
		}
	}
	return nil
}

// decompressIfNeeded wraps r with a zlib reader when the content is
// compressed. CMF byte 0x78 marks zlib (DEFLATE with a 32K window).
func decompressIfNeeded(r io.Reader) (io.Reader, error) {
	br := bufio.NewReader(r)
	head, err := br.Peek(1)
	if err != nil {
		if err == io.EOF {
			return br, nil
		}
		return nil, err
	}
	if head[0] == 0x78 {
		return zlib.NewReader(br)
	}
	return br, nil
}

// headerReader reads consecutive header fields with a sticky error, so the
// field sequence below stays readable.
type headerReader struct {
	d   *serdes.Deserializer
	err error
}

// u8, u32, and boolean currently have no call site that keeps their
// result — the header fields they read are skipped, not interpreted — but
// they return a value to stay symmetric with u16/gameVersion/modVersion/str,
// which do. Splitting the API into "value" and "skip" variants would read
// worse than the occasional discarded return.
//
//nolint:unparam
func (h *headerReader) u8() uint8 {
	var v uint8
	if h.err == nil {
		v, h.err = h.d.ReadU8()
	}
	return v
}

func (h *headerReader) u16() uint16 {
	var v uint16
	if h.err == nil {
		v, h.err = h.d.ReadU16()
	}
	return v
}

//nolint:unparam // see u8
func (h *headerReader) u32() uint32 {
	var v uint32
	if h.err == nil {
		v, h.err = h.d.ReadU32()
	}
	return v
}

func (h *headerReader) optimU32() uint32 {
	var v uint32
	if h.err == nil {
		v, h.err = h.d.ReadOptimU32()
	}
	return v
}

//nolint:unparam // see u8
func (h *headerReader) boolean() bool {
	var v bool
	if h.err == nil {
		v, h.err = h.d.ReadBool()
	}
	return v
}

func (h *headerReader) str() string {
	var v string
	if h.err == nil {
		v, h.err = h.d.ReadStr()
	}
	return v
}

func (h *headerReader) gameVersion() mod.GameVersion {
	var v mod.GameVersion
	if h.err == nil {
		v, h.err = h.d.ReadGameVersion()
	}
	return v
}

func (h *headerReader) modVersion() mod.MODVersion {
	var v mod.MODVersion
	if h.err == nil {
		v, h.err = h.d.ReadMODVersion()
	}
	return v
}

func parse(r io.Reader) (*File, error) {
	d := serdes.NewDeserializer(r)
	h := &headerReader{d: d}

	version := h.gameVersion()

	h.u8()         // unknown byte after the version
	h.str()        // campaign
	h.str()        // level_name
	h.str()        // base_mod
	h.u8()         // difficulty
	h.boolean()    // finished
	h.boolean()    // player_won
	h.str()        // next_level
	h.boolean()    // can_continue
	h.boolean()    // finished_but_continuing
	h.boolean()    // saving_replay
	h.boolean()    // allow_non_admin_debug_options
	h.modVersion() // loaded_from (MODVersion, not GameVersion)
	h.u16()        // loaded_from_build
	h.u8()         // allowed_commands
	h.boolean()    // unknown
	h.u32()        // unknown
	h.boolean()    // unknown

	modsCount := h.optimU32()
	var mods []MODEntry
	for range modsCount {
		name := h.str()
		modVersion := h.modVersion()
		h.u32() // CRC
		if h.err != nil {
			return nil, h.err
		}
		mods = append(mods, MODEntry{Name: name, Version: modVersion})
	}

	// Unknown 4 bytes between the MOD list and the startup settings.
	h.u32()
	if h.err != nil {
		return nil, h.err
	}

	startup, err := parseStartupSettings(d)
	if err != nil {
		return nil, err
	}

	return &File{Version: version, MODs: mods, StartupSettings: startup}, nil
}

func parseStartupSettings(d *serdes.Deserializer) (*settings.Section, error) {
	tree, err := d.ReadPropertyTree()
	if err != nil {
		return nil, err
	}

	section, err := settings.NewSection("startup")
	if err != nil {
		return nil, err
	}

	// Non-dictionary elements are skipped, as in the Ruby implementation:
	// each setting is expected to be a {"value": X} dictionary.
	entries, ok := tree.Dict()
	if !ok {
		return section, nil
	}
	for _, entry := range entries {
		wrapper, ok := entry.Value.Dict()
		if !ok {
			continue
		}
		for _, wrapperEntry := range wrapper {
			if wrapperEntry.Key == "value" {
				section.Set(entry.Key, wrapperEntry.Value)
				break
			}
		}
	}
	return section, nil
}
